Check links in the manual (#7541)
NixOS/nix: doc/manual/src/contributing/hacking.md

The installer now supports a `--tarball-url-prefix` flag which _may_ have
solved this need?
-->

### Checking the manual links

The build checks for broken internal links, but this happens late in the process,
so `nix build .` is not suitable for iterating. To check the manual incrementally, run:

```console
make html -j $NIX_BUILD_CORES
```

When iterating on the makefile, run:

```console
rm $(git ls-files doc/manual/ -o | grep -F '.md') && rmdir doc/manual/src/command-ref/new-cli && make html -j $NIX_BUILD_CORES
```

> If a broken link occurs in a snippet that was inserted into multiple generated files in different directories, use `@docroot@` to reference the `doc/manual/src` directory.

Reviewer: I'm not sure what this instruction is trying to tell me. In which case does this happen? I suppose you encountered this during debugging, but I'm lacking the context to follow. Maybe `@docroot@` should be documented in the Makefile itself (saying why it exists and when to use it)?

Author: It doesn't have a single place in the Makefile either, so this seems like a central place for it.

Author: I've linked this from the makefile in 4917ca17ce70c31ee5c0b16a707bfa5ec647f1e1

Reviewer: ```suggestion
```

Now that you have the sentence in the new subsection.

Reviewer: @roberth just this fix and you can merge yourself. I already approved.

This happens late in the process, so `nix build` is not suitable for iterating.
To build the manual incrementally, run:

```console
make html -j $NIX_BUILD_CORES
```

In order to reflect changes to the [Makefile], clear all generated files before re-building:

[Makefile]: https://github.com/NixOS/nix/blob/master/doc/manual/local.mk

```console
rm $(git ls-files doc/manual/ -o | grep -F '.md') && rmdir doc/manual/src/command-ref/new-cli && make html -j $NIX_BUILD_CORES
```

If a broken link occurs in a snippet that was inserted into multiple generated files in different directories, use `@docroot@` to reference the `doc/manual/src` directory.

[`mdbook-linkcheck`] does not implement checking [URI fragments] yet.

[`mdbook-linkcheck`]: https://github.com/Michael-F-Bryan/mdbook-linkcheck
[URI fragments]: https://en.m.wikipedia.org/wiki/URI_fragment

#### `@docroot@` variable

`@docroot@` provides a base path for links that occur in reusable snippets or other documentation that doesn't have a base path of its own.

If a broken link occurs in a snippet that was inserted into multiple generated files in different directories, use `@docroot@` to reference the `doc/manual/src` directory.

> If the `@docroot@` literal appears in an error message from the `mdbook-linkcheck` tool, the `@docroot@` replacement needs to be applied to the (probably) generated source file that mentions it. See existing `@docroot@` logic in the [Makefile]. Regular markdown files that are not copied to other markdown files do have a base path of their own and they can use relative paths instead of `@docroot@`.

Reviewer: ```suggestion
If the `@docroot@` literal appears in an error message from the `mdbook-linkcheck` tool, the `@docroot@` replacement needs to be applied to the generated source file that mentions it.
See existing `@docroot@` logic in the [Makefile].
Regular markdown files used for the manual have a base path of their own and they can use relative paths instead of `@docroot@`.
```
